### DISCUSSION TITLE
Chronological progress comments

### DISCUSSION THREAD

**Discussion by georgestevens99** [adamreed90/dapr-samples#6] [General]
5/5/22 1900 EDT.  I am adding the WebAPI service, WebAPIAhttp.  Just the standard VS template with no changes.
@adamreed90 @paulgyro 
Here are the settings I used in the VS WebAPI project setup.  Note NO SECURITY!
![image](https://user-images.githubusercontent.com/6517661/167043348-ec99d633-63ce-4ce2-b627-78e06f4b68ba.png)

5/5/2022 1952.  I have not yet commited my changes, but the WebAPIAhttp runs OK on my local dev system.
Here is how to run it on yours, and use a browser to verify it operates:

How to Run the WebAPIAhttp.
1) Pull the solution down to your system and build it.  It should build with no errors.
2) Open a Windows Terminal command prompt window.
3) Cd to  <YourRepoLocalPathName>\HelloDotNet6\WebAPIAhttp
4) Dotnet run
	a) It should run and say it is listening on  http://localhost:5129/
5) Open a browser (Chrome and Firefox worked for me) and put the following url in the navigation area:   
                                     http://localhost:5129/WeatherForecast
	a) This should display the weather forecast -- Sunny days ahead!
6) If you want to debug while using dotnet run just open the debug menu and select Attach.  Then scroll down and find the process named WebAPIAhttp.
	a) Then you can set some breakpoints, say in the controller, to see what there is to see.

1955  I will commit it now and submit a PR.  2005 Done. Ready for WebAPI development.
2015  Checked git and my system to make sure the PR was OK and my merge happened OK.  Now the master branch is completely up to date and ready to go.  I'm done for the night.

**Comment by iamadamreed** [comment]
Great! I have been swamped with work, so I hope to contribute a bunch this weekend!

**Reply by paulgyro** [reply]
Good job!  I'm traveling so it will be a few days before I can sync my head into this.  Certainly, something I think it going to be great for us and the community.

**Comment by georgestevens99** [comment]
5/7/22
I am working on the Console.  I am afraid I have made a mess of the repo!  It started with me deleting a merged branch on my dev system and went down hill from there.  I can still merge via a PR but everything is NOT being merged into master.  I need help.

**Reply by iamadamreed** [reply]
I definitely suggest giving GitKraken (https://www.gitkraken.com/) a go, it really makes Git life so much easier, especially when it comes to fixing mistakes, but no worries, I manage interns and their commits in my daily work and I'm the master of fixing mishaps, as long as you can get it pushed I can fix it on merge. I just made a latte about to get going on the gRPC Service myself! :D

**Reply by iamadamreed** [reply]
Also, I see you have been approving the PRs. Generally the idea is, you make commits to your branch til your feature/change set is complete, then submit a PR, then the team reviews it and approves it or makes suggestions. Once we get into more serious code commits we should stick to that format.

**Reply by georgestevens99** [reply]
I agree.   I was experimenting with the VS Git Menu, new in VS 2022.  It does all that is necessary.  My problem is that I do not understand how to use it properly.  Any was, it is a good idea to review PRs.  And I have one awaiting approval.  Thanks.  When you fix the stuff I have messed up, please let me know what I did wrong.  Thanks.

**Comment by iamadamreed** [comment]
After reviewing my `docker-compose` file that I had mentioned I had issues with dots in project names, it turns out it was a user error, I also consulted some other more knowledgeable Docker users and they said that the dots should not pose an issue. Therefore I have returned to the preferred naming convention of `Solution.ProjectName`.  I believe this provides a cleaner and easier to understand project structure, open to comment on this.

While re-scaffolding the gRPC API, I have implemented a common Data library ([N-Tier Service Design](https://docs.microsoft.com/en-us/azure/architecture/guide/architecture-styles/n-tier) / [N-Tier Application Design](https://en.wikipedia.org/wiki/Multitier_architecture)), where I have located the common Protos, these can be then referenced in the projects with a simple modification which I have illustrated in the project files. I will also explain how it works below.  You can also check out this project for a simple example (https://github.com/grpc/grpc-dotnet/tree/master/examples/Liber)

See the following project layout:
![image](https://user-images.githubusercontent.com/16042069/167265087-16d0f772-11da-4ccb-a7da-f386ee8bdeee.png)

Be sure to update the `option csharp_namespace` in the proto files to match that of the Data library: 
![image](https://user-images.githubusercontent.com/16042069/167265189-a289e001-b53d-435c-a1be-d605ba41a8e0.png)

The Proto files live inside the Data Layer of the application, and are consumed by the Services in the API Project, to include them in the API Project (or any other), simply add the following: 
![image](https://user-images.githubusercontent.com/16042069/167265111-15800ce6-a51e-477a-94e7-3c1f4ed663a4.png)

Now all referenced Protos may be kept in a single location for consumption throughout the application :D

**Reply by georgestevens99** [reply]
Great to hear you got past the dots.  It's great to see we can put protos together.  In service oriented apps the protos typically are considered DTOs and often get put in one or more Contracts libraries so that they can be used by multiple services and clients.  In other words they at a lower level in the dependency tree that services and clients.  

In service orientation there is an emphasis on decoupling various layers and this is often done by making DTOs separate from other business objects used within services.  So one of the first things a service does with a request is to use something like Automapper to map the incomming DTOs to business objects.  Jimmy Bogard has a number of good articles on DTOs vs objects and was the author of Automapper.  But that is for full production situations.  What you are doing with it in the gRPC service looks OK for us.

**Comment by georgestevens99** [comment]
5/7/22
I got the console talking to WebAPIAhttp/weatherforecast working.  It just gets a json object from that route on the service.

Mainly it 1) gives you a fairly simple framework to write your own commands with, and 2) shows the general format of how to code the console app for testing your service ops.  The only thing of major difference you will have to do with your code is use post rather than get as does weather forecast.  Please let me know if you have any questions or concerns.

As I showed above.  Then start the console and enter the UI command   'weather ttt 1 1' and it should run OK.  At least it did for me.

I have submitted a PR for this code. So once that is approved give it a try and let me know if it works for you, or needs fixing.  

I will be doing other things for the rest of the day, but will check my email today and tomorrow so as to prevent anyone from getting stuck.

**Comment by iamadamreed** [comment]
See branch `hellodotnet6-grpcapi` 

I have completed a working example of a ASP.NET MVC Website consuming a gRPC Service via Dapr with a shared common Proto.
PR: https://github.com/adamreed90/dapr-samples/pull/14

**Comment by iamadamreed** [comment]
I have began a Shopping Service example using Entity Framework (with MariaDb), and using the Redis State Store, I thought this would be a nice addition. See `hellodotnet6-shoppingservice` branch. 

I thought a more in depth example past the Greeter service would be helpful. What do you think?

@georgestevens99 @paulgyro

**Reply by georgestevens99** [reply]
What do you think? 

I think you need to seek the input of the Dapr Maintainers in charge of dapr/samples.  Ultimately they decide on the content (and its format and documentation) of everything that gets put in dapr/samples.  Further to save you from doing work they do not want, I suggest you do this sooner rather than later.  You can do this simply by submitting a Proposal issue in dapr/samples, telling them what you have and ask for their guidance.  That will start the conversation.

Further understand that the more "features" your code has, the more documentation you will have to write in your ReadMe docs saying how to use the features from a developer's perspective.  Please take a look at these sorts of docs for other projects in dapr/samples to  get an idea of the level of documentation required for dapr/samples so as to give you an idea of how many hours writing documentation will be required for your code.

That aside, congratulations on your new additions to the repo!  I am maxed out this week so it may be a few days before I have the time to look at them, but I will.  I am looking forward to that.

I hope you take the next step of opening a proposal in dapr/samples.  It can be just a really simple one.  That will serve to get a good conversation going!

Thanks.

**Comment by georgestevens99** [comment]
5/13/22 Did a code walk though, plus ran the grpc greeter service and the MVC web app.  Found bugs.  Documented all of that in the comment section of the PR.  Only a few minor issues remain.
